app/routes/docs.$lang.$ref.tsx

```diff
@@ -755,10 +756,47 @@ function InnerContainer({ children }: { children: React.ReactNode }) {
   );
 }
 
-function useDoc(): Doc | null {
-  let data = useMatches().slice(-1)[0].data;
-  if (!data) return null;
-  // @ts-expect-error -- useMatches types changed to `unknown`, need to validate
+const docValidator = () => {
+  const menuDocAttributesSchema = z
+    .object({
+      title: z.string(),
+      order: z.number().optional(),
+      new: z.boolean().optional(),
+    })
+    .and(z.record(z.string(), z.any()));
+
+  const menuDocSchema: z.ZodType<MenuDoc> = z.object({
+    attrs: menuDocAttributesSchema,
+    filename: z.string(),
+    children: z.array(z.lazy(() => menuDocSchema)),
+    slug: z.string(),
+    hasContent: z.boolean(),
+  });
+
+  const docSchema = z.object({
+    doc: z.object({
+      attrs: menuDocAttributesSchema,
+      filename: z.string(),
+      slug: z.string(),
+      children: z.array(menuDocSchema),
+      html: z.string(),
+      headings: z.array(
+        z
+          .object({
+            html: z.string().nullable(),
+            slug: z.string().optional(),
+          })
+          .transform((v) => ({ ...v, slug: v.slug })),
+      ),
+    }),
+  });
```

Comment:
Thanks for working on this, this is pretty robust checking. Again, I want to avoid sending `zod` to the client, but I also kind of want to avoid doing a bunch of client side logic to parse and validate every field on the doc.

I think we can just do some inferring and create a little type guard here:

```
function hasDoc(data: unknown): data is { doc: Doc } {
  return !!data && typeof data === "object" && "doc" in data;
}

function useDoc(): Doc {
  const data = useMatches().at(-1)?.data;
  invariant(hasDoc(data), "No doc data found in loader data");
  return data.doc;
}
```

I'm open to alternative approaches, basically I just don't think there's a big reason to parse the entire data set each time when we can pretty safely assume if there is a `doc` field it's going to be the appropriate type